internal/zinc-persist/src/main/java/xsbti/compile/analysis/GenericMapper.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Zinc - The incremental compiler for Scala.
+ * Copyright 2011 - 2017, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * This software is released under the terms written in LICENSE.
+ */
+
+package xsbti.compile.analysis;
+
+import xsbti.compile.MiniSetup;
+
+import java.io.File;
+
+
+/**
+ * Defines a generic interface to map the values of the analysis file 1-to-1.
+ */
+public interface GenericMapper {
```

Comment:
I think that having this distinction at the type level is important to avoid "mistakes" by the users. A good example of implementations that differ is [this one]() where the relative read mapper and the relative write mapper differ.
  
Have you tried defining your portable analysis mappers like this?
```scala
class PortableMapper(mapper: File => File) extends ReadMapper with WriteMapper {
  def mapSourceFile(x: File): File = mapper(x)
  def mapBinaryFile(x: File): File = mapper(x)
  def mapProductFile(x: File): File = mapper(x)
  def mapOutputDir(x: File): File = mapper(x)
  def mapSourceDir(x: File): File = mapper(x)
  def mapClasspathEntry(x: File): File = mapper(x)

  // TODO: Determine whether the rest of these need to be overridden in practice.
  def mapJavacOption(x: String): String = x
  def mapScalacOption(x: String): String = x
  def mapBinaryStamp(f: File, x: Stamp): Stamp = x
  def mapSourceStamp(f: File, x: Stamp): Stamp = x
  def mapProductStamp(f: File, x: Stamp): Stamp = x
  def mapMiniSetup(x: MiniSetup): MiniSetup = x
}
```
  
I think this way is much nicer. WDYT?